src/Link.cc

```diff
@@ -120,16 +93,215 @@ namespace kickcat
+        int32_t const datagrams = frame_nominal_.datagramCounter();
+
+        bool is_frame_sent_nominal = true;
+        frame_nominal_.setSourceMAC(PRIMARY_IF_MAC);
+        int32_t toWrite = frame_nominal_.finalize();
+        int32_t written = socket_nominal_->write(frame_nominal_.data(), toWrite);
+        if (written < 0)
+        {
+            is_frame_sent_nominal = false;
+            DEBUG_PRINT("Nominal: write failed\n");
+        }
+        else if (written != toWrite)
+        {
+            is_frame_sent_nominal = false;
+            DEBUG_PRINT("Nominal: Wrong number of bytes written");
+        }
+
+        bool is_frame_sent_redundancy = true;
+        frame_nominal_.setSourceMAC(SECONDARY_IF_MAC);
+        written = socket_redundancy_->write(frame_nominal_.data(), toWrite);
+        frame_nominal_.clear();
+        frame_redundancy_.clear();
+        frame_redundancy_.resetContext();
+
+        if (written < 0)
+        {
+            is_frame_sent_redundancy = false;
+            DEBUG_PRINT("Redundancy: write failed\n");
+        }
+
+        else if (written != toWrite)
+        {
+            is_frame_sent_redundancy = false;
+            DEBUG_PRINT("Redundancy: Wrong number of bytes written");
+        }
+
+        if (is_frame_sent_nominal or is_frame_sent_redundancy)
+        {
+            sent_frame_++;
+        }
+        else
+        {
+            for (int32_t i = 0; i < datagrams; ++i)
+            {
+                int32_t index = index_head_ - i - 1;
+                callbacks_[index].status = DatagramState::SEND_ERROR;
+            }
+        }
+    }
+
+    void Link::read()
+    {
+        if (readFrame(socket_redundancy_, frame_nominal_) < 0)
+        {
+            DEBUG_PRINT("Nominal read fail\n");
+        }
+
+        if (readFrame(socket_nominal_, frame_redundancy_) < 0)
+        {
+            DEBUG_PRINT("redundancy read fail\n");
+        }
+    }
+
+
+    void Link::addDatagramToFrame(uint8_t index, enum Command command, uint32_t address, void const* data, uint16_t data_size)
+    {
+        frame_nominal_.addDatagram(index, command, address, data, data_size);
+    }
+
+
+    void Link::resetFrameContext()
+    {
+        frame_nominal_.resetContext();
+        frame_redundancy_.resetContext();
+    }
+
+
+    bool Link::isDatagramAvailable()
+    {
+        return frame_nominal_.isDatagramAvailable() or frame_redundancy_.isDatagramAvailable();
+    }
+
+
+    std::tuple<DatagramHeader const*, uint8_t*, uint16_t> Link::nextDatagram()
+    {
+        bool nom = frame_nominal_.isDatagramAvailable();
+        bool red = frame_redundancy_.isDatagramAvailable();
+
+        auto [header_nominal, data_nominal, wkc_nominal] = frame_nominal_.nextDatagram();
+        auto [header_redundancy, data_redundancy, wkc_redundancy] = frame_redundancy_.nextDatagram();
+
+        // When using the bus without redundancy not nom and red is the used case.
```

Comment:
Why is this the case without redundancy? Isn't red the one without datagrams in this case?

Reply:
There is some ambiguity with the names frame_redundancy and frame_nominal.

When redundancy is working properly:
- nominal_frame is written on nominal_socket and read on redundancy_socket
- redundancy_frame is written on redundancy_socket and read on nominal socket
So the two frames are stored as member of Link under these names (point of view of the writing).

When there is no redundancy interface:
- nominal_frame is written on nominal_socket
- when reading redundancy socket there is nothing for nominal_frame (not connected)
- when reading nominal_socket the frame is stored in redundancy_frame.

Reply:
got it